### PR TITLE
Support zope.configuration 4.2+ and recent persistent

### PR DESCRIPTION
**Pull request by jamadden**
Use an 'OptionalDottedName' field for our ZCML 'id' type, because
that's what we really want, instead of `PythonIdentifier`.

If this still causes problems in the wild, this could go all the way
back to just a TextLine, but we really want the dots to enforce the
tree.

Also add 3.7 and drop 3.3 while here.

Tests will fail until zope.configuration 4.2.1 is released with
https://github.com/zopefoundation/zope.configuration/pull/37

### CODE REVIEW

**Comment by jamadden** [comment]
zope.configuration 4.2.1 was released and these builds are green.

File: .travis.yml [hunk 1]
```diff
@@ -2,11 +2,15 @@ language: python
 sudo: false
 python:
   - 2.7
-  - 3.3
   - 3.4
   - 3.5
   - 3.6
-  - pypy-5.4.1
+  - pypy
```
Comment (icemac) [inline .travis.yml:8]: Any chance to test pypy3, too?

File: CHANGES.rst [hunk 2]
```diff
@@ -2,11 +2,15 @@
  CHANGES
 =========
 
-4.0.1 (unreleased)
+4.1.0 (unreleased)
 ==================
 
-- Nothing changed yet.
+- Support newer zope.configuration and persistent. See `issue 2
+  <https://github.com/zopefoundation/zope.preference/issues/2>`_.
```
Comment (icemac) [inline CHANGES.rst:9]: Does it still run with older versions or should the minimum versions in `setup.py` changed to these newer versions?

File: src/zope/preference/README.rst [hunk 3]
```diff
@@ -234,9 +234,11 @@ that implements a default preferences provider that can be added as a unnamed
 utility for each site.
 
   >>> root
-  <zope.site.folder.Folder ...>
+  <...Folder ...>
```
Comment (icemac) [inline src/zope/preference/README.rst]: Maybe look at `root.__class__` instead for a more meaningful assertion.

**Review by icemac — APPROVED** [review]
LGTM besides some little questions.

File: .travis.yml (hunk 1 above)
Comment (jamadden) [reply .travis.yml:8]: Sure!

File: CHANGES.rst (hunk 2 above)
Comment (jamadden) [reply CHANGES.rst:9]: It should still run with older versions.

File: src/zope/preference/README.rst (hunk 3 above)
Comment (jamadden) [reply src/zope/preference/README.rst]: I found this part of the test disturbing and difficult to understand. I don't like it when variables just magically appear from thin air. The setup for `root` and `rsm` is very short, and there's only one doctest that uses it. What if we inline that setup here? Then the assertion isn't really needed at all.

Comment (icemac) [reply src/zope/preference/README.rst]: +1 to put the setup into the doctest.

**Pull request merged**